Component labeler

tmjd opened the pull request:
## Description

I wanted to add the new label (`app.kubernetes.io/name`) talked about in #1385 but didn't want to manually add it everywhere so I took a look at the Deployments and DaemonSets and found that they were very consistent and took this approach.
WDYT?


<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

caseydavenport commented on pkg/render/apiserver.go:
```diff
@@ -785,22 +785,21 @@ func (c *apiServerComponent) apiServerDeployment() *appsv1.Deployment {
 			Strategy: appsv1.DeploymentStrategy{
 				Type: appsv1.RecreateDeploymentStrategyType,
 			},
+			// Why do we use apiserver: true instead of k8s-app: name?
```
I doubt anyone knows at this point.

tmjd replied:
:grinning: I figured that was probably the case.

caseydavenport replied:
I think we should remove this comment, though. Or at least update it to say:

```
// For legacy reasons we use apiserver: true here instead of the k8s-app: name label, so we need to set it explicitly rather than use the common labeling logic. 
```

caseydavenport commented on pkg/controller/utils/component.go:
```diff
@@ -361,6 +363,48 @@ func ensureOSSchedulingRestrictions(obj client.Object, osType rmeta.OSType) {
+		if d.Spec.Selector == nil {
+			d.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"k8s-app": name,
```
Do we not want the app.kubernetes.io label here as well?

caseydavenport reviewed (commented):
My gut reaction is that maybe it's a little bit too magical, and we're still going to need a way for users to specify their own labels and annotations, which will need to be validated against the ones that the operator puts on by default.

But, I do like the general direction and a little bit of magic might be appropriate here (and some enforced consistency as well!) 

I think some comments explaining the pattern that we're enforcing would help a bit.

e.g., 

"For deployments, we ensure that they have labels X,Y,Z, that their pod selector is X, and that the pod template always has A,B,X."

caseydavenport commented:
@tmjd I was thinking a general code comment explaining what labels we apply to every Deployment / Daemonset, but maybe the code is self-explanatory. idk.

tmjd commented:
I'm not suggesting one is not needed, I'm just not sure where it would be useful. On the CreateOrUpdateOrDelete call?

caseydavenport commented:
I was imagining it would be at the top of `setCommonFields` (which maybe could be reduced in scope to "setStandardLabels" or something like that.

caseydavenport commented:
@tmjd should we bring this guy up to date?

tmjd commented:
I've updated this.
One notable change is I changed the setting of the labels to always be done by the setStandardSelectorAndLabels function, not only when the ObjectMeta.Labels was nil.

caseydavenport commented on pkg/controller/utils/component.go:
```diff
@@ -407,6 +410,51 @@ func ensureOSSchedulingRestrictions(obj client.Object, osType rmeta.OSType) {
+		if d.ObjectMeta.Labels == nil {
+			d.ObjectMeta.Labels = make(map[string]string)
+		}
+		d.ObjectMeta.Labels["k8s-app"] = name
```
Guess we need to keep this around for now, but wondering if we should eventually migrate off it to the app.kubernetes.io label and remove this altogether.

tmjd replied:
We certainly need to keep it until we switch over the matchSelector and I'm not sure if we can just switch that over on upgrade. I don't see much harm in keeping it.

caseydavenport reviewed (approved):
Couple minor comments, and a rebase, but otherwise lets get this in.
